### ISSUE TITLE
decoding json in swift

### ISSUE THREAD

**Issue by liquidoshin** [arackaf/my-blog#31]
Hello Adam. I'm assuming this repo was created for issues or questions found on your blog. Please direct me elsewhere if this is not the right place. I found a post on your blog online that was really close to answering my question but not entirely and I'm still having issues. And of course, with swift and any apple products there's never enough info online to really fully answer your issue. So I have some code that gets a json dictionary back from an in-house REST api but I keep getting an error when trying to decode that json into a swift struct. I've tried everything under the sun to get this to work and it's so seemingly a trivial issue you'd think I would have figured it out by now. The problem I'm experiencing is that the json dictionary being returned from the api has an array of strings as one of the items. For instance:

[{"hostid":123456,"machineid":123456,"hostname":"sample.host.name","ip":"0.0.0.0","mac":"aa:aa:aa:aa:aa:aa","owner":"123456","number":"1234567","admin":["123456"]}]

You'll notice the admin key is the one that has the array of strings as the value.

my struct I use to store this returned json dictionary is:

struct jsonDataStruct: Decodable {
    var hostid: Int = 0
    var machineid: Int = 0
    var hostname: String = ""
    var ip: String = ""
    var mac: String = ""
    var owner: String = ""
    var number: String = ""
    var admin: [String] = []
}
(I know you're not typically supposed to put initial values for the keys in the struct but because of the way I'm doing it in my code it gives me an error if I don't)

And my code handles this via a method that's called:

mutating func getJSONData() -> jsonDataStruct {
        var restapiData = jsonDataStruct()
        var exists = true
        let task = Process()
        let pipe = Pipe()
        task.launchPath = self.curl
        task.arguments = ["-k", "-s", "--request", "GET", "--negotiate", "--user", ":", "https://apiurl.com/......"]
        task.standardOutput = pipe
        task.launch()
        task.waitUntilExit()
        let outputData = pipe.fileHandleForReading.readDataToEndOfFile()
        let output = String(decoding: outputData, as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines)
        if output != "" {
            print("output from GET command: \(output)")
            let jsonData = output.data(using: .utf8)!
            restapiData = try! JSONDecoder().decode(jsonDataStruct.self, from: jsonData)
            print("hostmaster data in struct from REST api get call \(restapiData)")
            let outputArray = output.components(separatedBy: .newlines)
            for line in outputArray {
                if line.trimmingCharacters(in: .whitespacesAndNewlines).range(of: #"not found"#, options: .regularExpression) != nil {
                    exists = false
                    break
                }
            }
            if exists == true {
                print("hmData: \(hmData)")
                return hmData
            }
        }
        return hmData
    }

Calling this method and running the code inside gives me an error:
main/main.swift:394: Fatal error: 'try!' expression unexpectedly raised an error: Swift.DecodingError.typeMismatch(Swift.Dictionary<Swift.String, Any>, Swift.DecodingError.Context(codingPath: [], debugDescription: "Expected to decode Dictionary<String, Any> but found an array instead.", underlyingError: nil))
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.

It seems the swift code isn't handling the fact that there is an array of strings in the returned json even though I explicitly tell the struct that is storing this to store an array of strings for the admin key. Your blog has a similar example in it where it has an array of strings as one of the key's values in this specific topic https://adamrackis.dev/blog/swift-codable-any

But it didn't seem to fully touch on this issue. Could you possibly help me with this? You're my last hope lol. Thank you so much

**Comment by liquidoshin** [comment]
After writing all that I figured it out haha. The error wasn't complaining about the admin key and it containing an array of strings, it was that the returned data structure of the json was actually in an array and I needed to type cast the struct to an array:
restapiData = try! JSONDecoder().decode([jsonDataStruct].self, from: jsonData)

That fixed it. Sorry to take space on your git. Feel free to delete this if needed.
